ehrapy/api/preprocessing/_data_imputation.py

```diff
@@ -3,51 +3,300 @@
+
+    if isinstance(replacement, list):
+        column_indices = get_column_indices(adata, replacement)
+        adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+    # impute all columns if None passed
+    else:
+        adata.X = imputer.fit_transform(adata.X)
 
-    # scenario 2: Replace all missing values in a subset of columns with the specified value
-    # TODO
 
-    # scenario 3: Replace all missing values in a subset of columns with a specified value per column
-    # TODO
+def _miss_forest(adata: AnnData, **kwargs) -> AnnData:
+    """Impute data using the MissForest stratgey. See https://academic.oup.com/bioinformatics/article/28/1/112/219101.
+    This requires the computation of which columns in X contain numerical only (including NaNs) and which mixed data, which could be an
+    expensive operation on a larger X.
+    Speeding this up could be achieved by passing the "replacement" kwarg and indicate, which column is
+    numerical only and which mixed data.
+
+    Args:
+        adata: The AnnData object to use MissForest Imputation on
+        **kwargs: Keyword args
+                  replacement: A list of strings indicating which columns to use MissForest imputation on (if None -> all columns)
+
+    Returns:
+            The imputed (but unencoded) AnnData object
+    """
+    # TODO: passe replacement kwarg to indicate which column is numeric and which is mixed data (for those that should be imputed using MissForest strategy)
+    non_num_indices = _get_non_numerical_column_indices(adata.X)
+    num_indices = [ind for ind in range(adata.X.shape[1]) if ind not in non_num_indices]
+    non_num_indices = list(non_num_indices)
+    imp_num = IterativeImputer(estimator=RandomForestRegressor(), initial_strategy="mean", max_iter=1, random_state=0)
+    imp_cat = IterativeImputer(
+        estimator=RandomForestClassifier(), initial_strategy="most_frequent", max_iter=10, random_state=0
+    )
+
+    # encode all non numerical columns
+    enc = OrdinalEncoder()
+    adata.X[::, non_num_indices] = enc.fit_transform(adata.X[::, non_num_indices])
+    # perform the imputation strategy on both, numerical and non numerical data
+    adata.X[::, num_indices] = imp_num.fit_transform(adata.X[::, num_indices])
+    adata.X[::, non_num_indices] = imp_cat.fit_transform(adata.X[::, non_num_indices])
+    # decode ordinal encoding to obtain imputed original data
+    adata.X[::, non_num_indices] = enc.inverse_transform(adata.X[::, non_num_indices])
 
     return adata
+
+
+def _get_non_numerical_column_indices(X: np.ndarray) -> set:
+    """Return indices of columns, that contain at least one non numerical value that is not "Nan"."""
+    is_numeric_numpy = np.vectorize(_is_float, otypes=[bool])
+    mask = np.apply_along_axis(is_numeric_numpy, 0, X)
+
+    _, column_indices = np.where(~mask)
+    non_num_indices = set(column_indices)
+
+    return non_num_indices
+
+
+def _is_float(val):
```

Comment:
I have the impression that I've seen this function somewhere already. Maybe in `encode` or something?